source/core_mqtt_agent.c

```diff
@@ -621,9 +643,9 @@ static void handleAcks( const MQTTAgentContext_t * pAgentContext,
 
 static MQTTAgentContext_t * getAgentFromMQTTContext( MQTTContext_t * pMQTTContext )
 {
-    void * ret = pMQTTContext;
+    size_t offset = ( ( size_t ) &( ( ( MQTTAgentContext_t * ) 0 )->mqttContext ) );
```

Comment:
I see what you are doing here - cool way to portably determine the correct struct offset. im just worried that this will not be misra compliant - but could probably be fixed when we make this lib misra compliant at some point.

Reply:
The earlier was not only a violation to MISRA rules, it was also a possible memory corruption issue in case anyone decided to modify the MQTT-Agent structure.

Reply:
I'm pretty sure this now invokes undefined behavior though

Reply:
Also size_t is not guaranteed to hold a pointer converted to int (This is the case on at least one upstream FreeRTOS port and on CHERI archs).

Reply:
Maybe
```
static MQTTAgentContext_t * getAgentFromMQTTContext( MQTTContext_t * pMQTTContext )
{
    MQTTAgentContext_t  ctx = { 0 };
    ptrdiff_t offset = ( ( uint8_t * ) &( ctx->mqttContext ) ) - ( uint8_t * ) ctx;
    return ( MQTTAgentContext_t * ) &( ( ( uint8_t * ) pMQTTContext )[ - offset ] );
}
```